### PR TITLE
Add optional filter for feature flag in user api

### PR DESCRIPTION
**Pull request by akashkj**
## Product Description
Enhanced the API endpoint for getting  domain membership info for a user by providing additional filter for feature flag.

## Technical Summary
https://dimagi-dev.atlassian.net/browse/SC-2069

### Safety story
Performed local testing

### Automated test coverage
Added unit tests

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/api/resources/v0_5.py [hunk 1]
```diff
@@ -799,15 +799,20 @@ def dispatch_list(self, request, **kwargs):
...
 
-    def get_object_list(self, request):
+    def get_object_list(self, request, feature_flag=None):
         couch_user = CouchUser.from_django_user(request.user)
         results = []
         for domain in couch_user.get_domains():
             if not domain_has_privilege(domain, privileges.ZAPIER_INTEGRATION):
                 continue
             domain_object = Domain.get_by_name(domain)
+            if feature_flag and feature_flag not in toggles.toggles_dict(domain=domain).keys():
```
Comment (proteusvacuum) [inline corehq/apps/api/resources/v0_5.py:814]: Do you need to check if the FF is enabled here? Also, a user may have a flag enabled, but not a domain - what is the outcome there?

File: corehq/apps/api/resources/v0_5.py [hunk 2]
```diff
@@ -799,15 +799,20 @@ def dispatch_list(self, request, **kwargs):
                 raise
 
     def obj_get_list(self, bundle, **kwargs):
-        return self.get_object_list(bundle.request)
+        feature_flag = bundle.request.GET.get("feature_flag")
+        if feature_flag and feature_flag not in toggles.all_toggles_slug():
+            raise BadRequest("{} is not a valid feature flag".format(feature_flag))
```
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/api/resources/v0_5.py:804]: I didn't understood completely. `obj_get_list()` is getting called via API trigger and it internally calls `get_object_list()`. When  can we expect it to be called directly?

Modified string format.

[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/apps/api/resources/v0_5.py]: Ah. OK, then if `get_object_list()` should _not_ be called directly, then by Python convention it is marked as protected by prefixing it with an underscore. That tells developers / users of the class not to call that method from outside the class. e.g.
```python
    def _get_object_list(self, request, feature_flag=None):
        ...
```

But if `get_object_list()` _could_ be called directly, then it should be responsible for validating/cleaning its inputs.

File: corehq/apps/api/resources/v0_5.py (hunk 1 above)
Comment (kaapstorm) [reply corehq/apps/api/resources/v0_5.py]: > What is the expected behavior if flag enabled for user but not for domain?

Then it is enabled. i.e. It is an "or" operation: enabled = enabled_for_domain or enabled_for_user

> We may need to pass user also in the call along with domain accordingly.

:+1:

File: corehq/apps/api/resources/v0_5.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/api/resources/v0_5.py]: Got it. Will prefix the method.

**Review by sravfeyn — APPROVED** [review]
Thanks, look good. LGTM, if others' comments are addressed as well.

**Review by kaapstorm — APPROVED** [review]

File: corehq/toggles/__init__.py [hunk 3]
```diff
@@ -496,6 +496,13 @@ def all_toggles():
     return list(all_toggles_by_name().values())
 
 
+def all_toggles_slug():
```
Comment (proteusvacuum) [inline corehq/toggles/__init__.py:499]: nit: I think `all_toggle_slugs` is a more correct name for this

**Review by proteusvacuum — APPROVED** [review]

File: corehq/apps/api/resources/v0_5.py [hunk 4]
```diff
@@ -802,12 +802,17 @@ def obj_get_list(self, bundle, **kwargs):
...
+        feature_flag = request.GET.get("feature_flag")
+        if feature_flag and feature_flag not in toggles.all_toggle_slugs():
+            raise BadRequest(f"{feature_flag!r} is not a valid feature flag")
         couch_user = CouchUser.from_django_user(request.user)
         results = []
         for domain in couch_user.get_domains():
             if not domain_has_privilege(domain, privileges.ZAPIER_INTEGRATION):
                 continue
             domain_object = Domain.get_by_name(domain)
+            if feature_flag and feature_flag not in toggles.toggles_dict(username=request.user, domain=domain):
```
Comment (calellowitz) [inline corehq/apps/api/resources/v0_5.py:814]: This is causing errors. `request.user` is a user object, not a username string, which this function is expecting. https://sentry.io/organizations/dimagi/issues/3294412535/?query=is%3Aunresolved+bad+type&statsPeriod=14d

I think you just need `request.user.username`

Comment (akashkj) [reply corehq/apps/api/resources/v0_5.py:814]: Thanks @calellowitz. PR: https://github.com/dimagi/commcare-hq/pull/31662

**Pull request merged**